dispatcher/backend/src/periodic-tasks.py

```diff
@@ -39,113 +41,110 @@
 
     logger.info(f":: removing tasks history (>{HISTORY_TASK_PER_SCHEDULE})")
-    cursor = Tasks().aggregate(
-        [
-            {"$group": {"_id": "$schedule_name", "count": {"$sum": 1}}},
-            {"$match": {"count": {"$gt": HISTORY_TASK_PER_SCHEDULE}}},
-        ]
-    )
-
-    schedules_with_too_much_tasks = [s["_id"] for s in cursor]
-
-    task_ids_to_delete = []
-    for schedule_name in schedules_with_too_much_tasks:
-        cursor = Tasks().aggregate(
-            [
-                {"$match": {"schedule_name": schedule_name}},
-                {
-                    "$project": {
-                        "schedule_name": 1,
-                        "updated_at": {"$arrayElemAt": ["$events.timestamp", -1]},
-                    }
-                },
-                {"$sort": {"updated_at": pymongo.DESCENDING}},
-                {"$skip": HISTORY_TASK_PER_SCHEDULE},
-            ]
-        )
-        task_ids_to_delete += [t["_id"] for t in cursor]
-
-    result = Tasks().delete_many({"_id": {"$in": task_ids_to_delete}})
-    logger.info(f"::: deleted {result.deleted_count}/{len(task_ids_to_delete)} tasks")
 
+    schedule_ids_stmt = (
+        sa.select(dbm.Task.schedule_id)
+        .group_by(dbm.Task.schedule_id)
+        .having(sa.func.count(dbm.Task.id) > 10)
+    )
 
-def status_to_cancel(now, status, timeout):
+    schedules_with_too_much_tasks = session.execute(
+        sa.select(dbm.Schedule)
+        .filter(dbm.Schedule.id.in_(schedule_ids_stmt))
+        .group_by(dbm.Task.schedule_id)
+        .having(sa.func.count(dbm.Task.id) > HISTORY_TASK_PER_SCHEDULE)
+    ).scalars()
+
+    cnt_delete_requests = 0
+    for schedule in schedules_with_too_much_tasks:
+        cnt = 0
+        for task in sorted(schedule.tasks, lambda x: -x.updated_at):
+            if cnt < HISTORY_TASK_PER_SCHEDULE:
+                cnt += 1
+                continue
+            session.delete(task)
+            cnt_delete_requests += 1
+    logger.info(f"::: deleted {cnt_delete_requests} tasks")
+
+
+def status_to_cancel(now, status, timeout, session: so.Session):
     logger.info(f":: canceling tasks `{status}` for more than {timeout}s")
     ago = now - datetime.timedelta(seconds=timeout)
-    query = {"status": status, f"timestamp.{status}": {"$lte": ago}}
-    result = Tasks().update_many(
-        query,
-        {
-            "$set": {
-                "status": TaskStatus.canceled,
-                "canceled_by": NAME,
-                f"timestamp.{TaskStatus.canceled}": now,
-            },
-            "$push": {
-                "events": {
-                    "code": TaskStatus.canceled,
-                    "timestamp": now,
-                }
-            },
-        },
-    )
-    logger.info(f"::: canceled {result.modified_count}/{result.matched_count} tasks")
+    tasks = session.execute(
+        sa.select(dbm.Task)
+        .filter(dbm.Task.status == status)
+        .filter(dbm.Task.timestamp[status].astext <= ago)
```

Comment:
`astext` is a SQLAlchemy Postgresql operator for JSON : https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.JSON.Comparator.astext

It is used to tell SQLAlchemy that instead of using the "->" JSON operator, it must use the "->>" JSON operator (more details on these operators here : https://www.postgresql.org/docs/current/functions-json.html)

You need this to tell Postgresql engine that the result of JSON operation (lookup of key here) is indeed a scalar value and must be extracted so that it can be compared (otherwise it will fail since it does not know how to compare a result of type `json`).

Reply:
OK thanks. So the *text* here refers to the Key name which is a string and not the value.

Reply:
Here `dbm.Task.timestamp[status].astext` is a string which is the status of the timestamp while `dbm.Task.timestamp[status]` would be a JSON object which indeed is a string in our dataset but Postgresql don't know if this can be assumed and so don't know how to apply the `<=` comparison between a JSON object and the constant value `ago`